### DISCUSSION TITLE
Typing dynamic subclass

### DISCUSSION THREAD

**Discussion by d-tw** [microsoft/pyright#5222] [Q&A] [answered]
Hi,

This may not currently be possible, but I wanted to make sure I'm not missing anything.

I'm trying to type a class decorator that returns a new subclass with a mixin - in other words, I'm trying to merge the type signature of a class with a mixin. In Typescript, for class `C` and mixin `M`, the equivalent would be `C & M`.

The issue I'm currently facing is that my resulting type contains the methods from `M`, but loses the signature of `C`.

Is it possible to create this intersection type signature?

```py
from typing import TypeVar, Type

T = TypeVar('T', bound=object)

class JSONMixin():
    @classmethod
    def from_json(cls: Type[T], json: str) -> T:
        ...

def apply_mixin(cls: Type[T]):
    class B(JSONMixin, cls):
        ...

    return B

class A:
    val: str

    def __init__(self, val: str):
        self.val = val

A2 = apply_mixin(A)

a2_1 = A2.from_json('')
a2_2 = A2('val')

reveal_type(a2_1)
reveal_type(a2_1.val)
reveal_type(a2_2)
```

<img width="775" alt="image" src="https://github.com/microsoft/pyright/assets/3317531/8cc13d7c-2566-414b-86b0-1c0e74359e29">

**Comment by hmc-cs-mdrissi** [comment, marked as answer]
Here's the [issue](https://github.com/python/typing/issues/213) you are looking for. Intersections would be very nice for this exact use case of class decorators. It'd also allow type safe way to do dynamic patching of extra methods/attributes. Sadly I'm unaware of any good way to do this today in current type system.

Closest I can offer is make from_json a top level function and have class decorator register cls so that it can be used with it. Something like,

```python
def from_json(cls: Type[T], json: str) -> T:
  ...

def apply_mixin(cls: Type[T]) -> type[T]:
  register_json(cls)
  return cls

@apply_mixin
class A:
  def __init__(self, val: str):
        self.val = val


a2_1 = from_json(A, '')
a2_2 = A("val")
```

I do have actual code I use heavily in my codebase that follows this pattern. I don't see a good way to add classmethod to A type safe way, but you can still have your decorator register json support and then call it later.

This does leave issue that you can write code like,

```python
class B:
  ...

from_json(B, "...") # Crashes as apply_mixin was never done to B
```

I don't see how to make from_json method that allows registration with decorator and checks type passed was actually registered. If you have control of full class hierarchy then you can do,

```python
class C(JsonMixin):
  ...
```

directly, but if you want to work with external classes and just call function to add mixin I don't see how.

**Reply by d-tw** [reply]
Hi,

Thanks for the quick reply - this confirms what I thought.

Unfortunately, I can't refactor the code, as I'm trying to provide some minimal type hints to a 3rd party lib.
